[libbeat] kafka message support headers feature

herbxu opened the pull request:
<!-- Type of change
Please label this PR with one of the following labels, depending on the scope of your change:
- Enhancement
-->

## What does this PR do?
kafka message support headers feature

## Why is it important?
Kafka Message Header lets consumers process message metadata without having to deserialize the message value/payload. This could be useful in the routing of message use case.

## Checklist
- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have made corresponding change to the default configuration files
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.

## Author's Checklist
xiangwei.xxw@alibaba-inc.com

<!-- Recommended
Add a checklist of things that are required to be reviewed in order to have the PR approved
-->
- [ ]

## How to test this PR locally

<!-- Recommended
Explain here how this PR will be tested by the reviewer: commands, dependencies, steps, etc.
-->

## Related issues

<!-- Recommended
Link related issues below. Insert the issue link or reference after the word "Closes" if merging this should automatically close it.

- Closes #123
- Relates #123
- Requires #123
- Superseds #123
-->
- 

## Use cases

<!-- Recommended
Explain here the different behaviors that this PR introduces or modifies in this project, user roles, environment configuration, etc.

If you are familiar with Gherkin test scenarios, we recommend its usage: https://cucumber.io/docs/gherkin/reference/
-->

## Screenshots

<!-- Optional
Add here screenshots about how the project will be changed after the PR is applied. They could be related to web pages, terminal, etc, or any other image you consider important to be shared with the team.
-->

## Logs

<!-- Recommended
Paste here output logs discovered while creating this PR, such as stack traces or integration logs, or any other output you consider important to be shared with the team.
-->

elasticmachine commented:
## :grey_exclamation: Build Aborted
  
  > The PR is not allowed to run in the CI yet
  


<!-- BUILD BADGES-->
> _the below badges are clickable and redirect to their specific view in the CI or DOCS_
[![Pipeline View](https://img.shields.io/badge/pipeline-pipeline%20-green)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats%2FPR-29940/detail/PR-29940/1//pipeline) [![Test View](https://img.shields.io/badge/test-test-green)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats%2FPR-29940/detail/PR-29940/1//tests) [![Changes](https://img.shields.io/badge/changes-changes-green)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats%2FPR-29940/detail/PR-29940/1//changes) [![Artifacts](https://img.shields.io/badge/artifacts-artifacts-yellow)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats%2FPR-29940/detail/PR-29940/1//artifacts)   [![preview](https://img.shields.io/badge/elastic-observability-blue)](https://ci-stats.elastic.co/app/apm/services/beats-ci/transactions/view?rangeFrom=2022-01-21T05:07:35.911Z&rangeTo=2022-01-21T05:27:35.911Z&transactionName=Beats%2Fbeats%2FPR-%7Bnumber%7D&transactionType=unknown&latencyAggregationType=avg&traceId=8cb6f7caafa82b55e735d0ad96b1bd76&transactionId=3b582debf67f973a)

<!-- BUILD SUMMARY-->
<details><summary>Expand to view the summary</summary>
<p>

#### Build stats



* Reason: `The PR is not allowed to run in the CI yet`


* Start Time: 2022-01-21T05:17:35.911+0000


* Duration: 5 min 41 sec


* Commit: c8b6cc795ab4fce5a86f1dc2436152084901cabe




</p>
</details>

<!-- TEST RESULTS IF ANY-->


<!-- STEPS ERRORS IF ANY -->












## :robot: GitHub comments

To re-run your PR in the CI, just comment with:



- `/test` : Re-trigger the build.


- `/package` : Generate the packages and run the E2E tests.


- `/beats-tester` : Run the installation tests with beats-tester.


- `run` `elasticsearch-ci/docs` : Re-trigger the docs validation. (use unformatted text in the comment!)




<!--COMMENT_GENERATED_WITH_ID_comment.id-->

elasticmachine commented:
Pinging @elastic/elastic-agent-data-plane  (Team:Elastic-Agent-Data-Plane)

rdner commented on libbeat/outputs/kafka/client.go:
```diff
@@ -76,7 +78,8 @@ func newKafkaClient(
 	topic outil.Selector,
-	writer codec.Codec,
+	headers map[string]string,
+    writer codec.Codec,
```
could you format the whitespace, please?

herbxu replied:
I have formatted this code, thanks !

herbxu replied:
> could you format the whitespace, please?

I have formatted this code, thanks !

rdner reviewed (commented):
Thank you for your contribution, @herbxu!

In order to merge this PR, we would need a test confirming the new functionality works (see https://github.com/elastic/beats/blob/main/libbeat/outputs/kafka/kafka_integration_test.go). Once it's there, I'm happy to merge the PR.

herbxu commented:
> Thank you for your contribution, @herbxu!
> 
> In order to merge this PR, we would need a test confirming the new functionality works (see https://github.com/elastic/beats/blob/main/libbeat/outputs/kafka/kafka_integration_test.go). Once it's there, I'm happy to merge the PR.

I have added the integration test for this feature, thanks !

rdner reviewed (commented):
Thank you for addressing the comments and adding the test case. I noticed another small improvement that you might want to apply.

It looks like you have not signed the CLA https://www.elastic.co/contributor-agreement

Since you use the privacy feature on GitHub and your commits have a different email address, the automation is not able to detect the CLA signature. However, I checked manually in our contributor database and I could find neither your nickname `herbxu` nor your email address xiangwei.xxw@alibaba-inc.com

After the CLA is signed I'll run the CI pipeline and we can merge this PR.

herbxu commented:
> 



> Thank you for addressing the comments and adding the test case. I noticed another small improvement that you might want to apply.
> 
> It looks like you have not signed the CLA https://www.elastic.co/contributor-agreement
> 
> Since you use the privacy feature on GitHub and your commits have a different email address, the automation is not able to detect the CLA signature. However, I checked manually in our contributor database and I could find neither your nickname `herbxu` nor your email address [xiangwei.xxw@alibaba-inc.com](mailto:xiangwei.xxw@alibaba-inc.com)
> 
> After the CLA is signed I'll run the CI pipeline and we can merge this PR.

I have signed with the email [xiangwei.xxw@alibaba-inc.com](mailto:xiangwei.xxw@alibaba-inc.com), thanks for your suggestion.

rdner commented:
Jenkins, please test this.

rdner commented:
/test

rdner commented:
@herbxu just FYI, I had to modify the configuration structure a little bit, see https://github.com/elastic/beats/pull/30369
